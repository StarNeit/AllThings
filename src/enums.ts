// Locales
export enum Locale {
  de_DE = 'de_DE',
  en_US = 'en_US',
  fr_FR = 'fr_FR',
  it_IT = 'it_IT',
  nl_NL = 'nl_NL',
  pt_PT = 'pt_PT',
}

// MicroApps
export enum MicroApps {
  ARCHILOGIC = 'archilogic',
  BOOKING = 'booking',
  CONCIERGE = 'concierge',
  CLIPBOARD = 'clipboard',
  COBOT = 'cobot',
  CONSUMPTION = 'consumption',
  CRAFTSPEOPLE = 'craftspeople',
  DOCUMENTS = 'documents',
  E_CONCIERGE = 'e-concierge',
  EXTERNAL_CONTENT = 'external-content',
  EXTERNAL_LINK = 'external-link',
  FALLBACK = '__fallback__',
  HELPDESK = 'helpdesk',
  MARKETPLACE = 'marketplace',
  MY_CONTRACTS = 'my-contracts',
  PINBOARD = 'community-articles',
  PROJECT = 'project',
  SETTINGS = 'settings',
  SHARING = 'sharing',
  UNIT = 'unit',
  WHO_IS_WHO = 'who-is-who',
}

// Ticket statuses.
export enum TicketStatus {
  CLOSED = 'closed',
  WAITING_FOR_AGENT = 'waiting-for-agent',
  WAITING_FOR_CUSTOMER = 'waiting-for-customer',
  WAITING_FOR_EXTERNAL = 'waiting-for-external',
}

// Ticket conversation message types.
export enum ConversationMessageType {
  FILE = 'file',
  TEXT = 'text',
}

// Payment statuses.
export enum PaymentStatus {
  SUCCESS = 'success',
  FAILURE = 'failure',
  PENDING_MANUAL = 'pending_manual',
  PENDING_STRIPE = 'pending_stripe',
  CANCELLED = 'cancelled',
  ABANDONED = 'abandoned',
  MANUAL_ACTION_NEEDED = 'manual_action_needed',
}
