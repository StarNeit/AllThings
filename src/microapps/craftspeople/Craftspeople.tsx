import {
  ChevronRightListItem,
  GroupedCardList,
  HorizontalView,
  SimpleLayout,
} from '@allthings/elements'
import Microapp from 'components/Microapp'
import withMixpanel from 'containers/withMixpanel'
import withRequest, { IWithRequest } from 'containers/withRequest'
import { MicroApps } from 'enums'
import { css } from 'glamor'
import React from 'react'
import { connect } from 'react-redux'
import { IAddress } from 'store/reducers/authentication'
import CraftspeopleList from './CraftspeopleList'
import MicroappBigTitleBar from 'components/TitleBar/MicroappBigTitleBar'
import GenericBackTitleBar from 'components/TitleBar/GenericBackTitleBar'
import { Locale } from '../../enums'

interface IProps {
  readonly addresses: IAddress[]
  readonly locale: string
  readonly introText?: IndexSignature<string>
  readonly mixpanel: (eventName: string) => void
}

export interface ICategory {
  readonly description: Partial<Record<Locale, string>>
  readonly id: string
  readonly key: string
  readonly name: Partial<Record<Locale, string>>
  readonly thirdPartyData: boolean
  readonly vocabulary: string
}

interface ICraftspersonEmbedded {
  readonly externalAgentCompany?: {
    readonly _embedded?: {
      readonly logo?: string
    }
  }
}

export interface ICraftsperson {
  readonly _embedded: ICraftspersonEmbedded
  readonly categories: ICategory[]
  readonly companyAddress: {
    city: string
    houseNumber: string
    postalCode: string
    street: string
  }
  readonly companyId: string
  readonly companyName: string
  readonly companyPhoneNumber: string
  readonly id: string
  readonly phoneNumber: string
  readonly username: string
}

interface IState {
  readonly addressId: string
  readonly craftspeople: ICraftsperson[]
  readonly loading: boolean
}

class Craftspeople extends React.PureComponent<IProps & IWithRequest, IState> {
  state = {
    addressId: '',
    craftspeople: [] as ICraftsperson[],
    loading: true,
  }

  async componentDidMount() {
    const { addresses, mixpanel } = this.props

    mixpanel('craftsmen.overview.viewed')

    addresses.length === 1
      ? await this.updateAddressCraftspeople(addresses[0].id)
      : this.setState({ loading: false })
  }

  handleBack = () => this.setState({ addressId: '' })

  updateAddressCraftspeople = async (id: string) => {
    const { entity } = await this.props.createRequest({
      method: 'GET',
      path: `/api/v1/utilisation-periods/${id}/craftspeople`,
    })

    this.setState({
      addressId: id,
      craftspeople: entity._embedded.items,
      loading: false,
    })
  }

  renderAddressList = () =>
    this.props.addresses.map(({ id, key }) => (
      <ChevronRightListItem
        data-e2e={`${id}-craftspeople-address`}
        key={key}
        onClick={() => this.updateAddressCraftspeople(id)}
      >
        {key}
      </ChevronRightListItem>
    ))

  renderTitleBar = (onlyOneAddress: boolean) => {
    const { addressId } = this.state

    return (
      <>
        {onlyOneAddress || (!onlyOneAddress && !addressId) ? (
          <MicroappBigTitleBar type={MicroApps.CRAFTSPEOPLE} />
        ) : (
          <GenericBackTitleBar onBack={this.handleBack} />
        )}
      </>
    )
  }

  render() {
    const { addresses, locale, introText } = this.props
    const onlyOneAddress = addresses.length === 1
    const { addressId, craftspeople, loading } = this.state

    return (
      <Microapp>
        <SimpleLayout padded={'horizontal'}>
          {this.renderTitleBar(onlyOneAddress)}
          <HorizontalView {...css({ height: '100%' })}>
            {!onlyOneAddress && (
              <GroupedCardList title="">
                {this.renderAddressList()}
              </GroupedCardList>
            )}
            {(addressId || onlyOneAddress) && (
              <CraftspeopleList
                addressId={addressId}
                craftspeople={craftspeople}
                introText={introText}
                loading={loading}
                locale={locale}
              />
            )}
          </HorizontalView>
        </SimpleLayout>
      </Microapp>
    )
  }
}

const mapStateToProps = ({ authentication: { user }, app }: IReduxState) => ({
  addresses: user.addresses,
  introText: app.microApps.find(
    microApp => microApp.type === MicroApps.CRAFTSPEOPLE,
  )?.introText,
  locale: user.locale,
})

export default connect(
  mapStateToProps,
  null,
)(withRequest(withMixpanel(Craftspeople)))
