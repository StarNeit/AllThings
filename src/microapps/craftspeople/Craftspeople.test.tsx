import React from 'react'
import { Provider } from 'react-redux'
import Craftspeople from './Craftspeople'
import { MicroApps } from 'enums'

const storeBase = {
  app: {
    config: {
      clientID: 123,
    },
    embeddedLayout: false,
    microApps: [
      {
        customLogo: 'someLogo',
        type: MicroApps.CRAFTSPEOPLE,
        label: { en_US: 'Craftspeople' },
      },
    ],
  },
}

const ADDRESS1 = '[data-e2e="1234-craftspeople-address"]'
const ADDRESS2 = '[data-e2e="3456-craftspeople-address"]'

describe('Craftspeople', () => {
  it('should render Craftspeople', async () => {
    const store = global.mockStore({
      ...storeBase,
      authentication: {
        accessToken: 'abcd1234',
        user: {
          addresses: [
            { id: '1234', key: 'Address#1' },
            { id: '3456', key: 'Address#2' },
          ],
          locale: 'en_US',
        },
      },
    })
    const wrapper = global.mountIntl(
      global
        .shallow(
          <Provider store={store}>
            <Craftspeople />
          </Provider>,
        )
        .get(0),
    )

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find(ADDRESS1).exists()).toBeTruthy()
    expect(wrapper.find(ADDRESS2).exists()).toBeTruthy()
  })
})
