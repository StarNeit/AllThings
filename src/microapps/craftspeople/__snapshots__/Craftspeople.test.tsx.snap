// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Craftspeople should render Craftspeople 1`] = `
.css-0,
[data-css-0] {
  width: 100%;
  background: white;
  box-shadow: 2px 2px 2px rgba(0, 0, 0, 0.05);
}

.css-1,
[data-css-1] {
  width: 12px;
  height: 12px;
}

.css-2,
[data-css-2] {
  width: 100%;
}

.css-3,
[data-css-3] {
  min-height: 0;
  background-color: #f3f5f7;
}

.css-4,
[data-css-4] {
  font-size: 42px;
  font-weight: bold;
  line-height: 1em;
  margin-bottom: 7px;
  letter-spacing: -2px;
  word-wrap: break-word;
}

.css-5,
[data-css-5] {
  height: 100%;
}

.css-6,
[data-css-6] {
  box-sizing: border-box;
  align-content: stretch;
  align-items: stretch;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  flex: 1;
  -webkit-align-content: stretch;
  -webkit-box-align: stretch;
  -webkit-align-items: stretch;
  -webkit-box-orient: vertical;
  -webkit-box-direction: normal;
  -webkit-flex-direction: column;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -webkit-flex: 1;
}

.css-7,
[data-css-7] {
  -webkit-overflow-scrolling: touch;
  overflow: auto;
  overflow-x: hidden;
  background-color: #f3f5f7;
  height: 100%;
}

.css-8,
[data-css-8] {
  min-width: 0;
  width: 100%;
  padding: 20px 15px 0px;
}

.css-9,
[data-css-9] {
  box-sizing: border-box;
  align-content: stretch;
  align-items: stretch;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  flex: 0 0 auto;
  -webkit-align-content: stretch;
  -webkit-box-align: stretch;
  -webkit-align-items: stretch;
  -webkit-box-orient: horizontal;
  -webkit-box-direction: normal;
  -webkit-flex-direction: row;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -webkit-flex: 0 0 auto;
}

.css-10,
[data-css-10] {
  box-sizing: border-box;
  flex: 1 1 100%;
  -webkit-flex: 1 1 100%;
}

.css-11,
[data-css-11] {
  box-sizing: border-box;
  flex: 0 0 auto;
  -webkit-flex: 0 0 auto;
}

.css-12,
[data-css-12] {
  display: inline-block;
}

.css-13,
[data-css-13] {
  overflow: hidden;
}

.css-14,
[data-css-14] {
  transform: translate3d(0%, 0, 0);
  transition: .5s;
  -webkit-transform: translate3d(0%, 0, 0);
  -webkit-transition: .5s;
  -moz-transition: .5s;
}

.css-15,
[data-css-15] {
  box-sizing: border-box;
  flex: 1;
  -webkit-flex: 1;
}

.css-16,
[data-css-16] {
  box-sizing: border-box;
  align-content: stretch;
  align-items: stretch;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  flex: 1;
  -webkit-align-content: stretch;
  -webkit-box-align: stretch;
  -webkit-align-items: stretch;
  -webkit-box-orient: horizontal;
  -webkit-box-direction: normal;
  -webkit-flex-direction: row;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -webkit-flex: 1;
}

.css-17,
[data-css-17] {
  box-sizing: border-box;
  align-content: stretch;
  align-items: stretch;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  flex: 0 0 auto;
  -webkit-align-content: stretch;
  -webkit-box-align: stretch;
  -webkit-align-items: stretch;
  -webkit-box-orient: vertical;
  -webkit-box-direction: normal;
  -webkit-flex-direction: column;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -webkit-flex: 0 0 auto;
}

.css-18,
[data-css-18] {
  padding-left: 15px;
  padding-right: 15px;
  padding-top: 15px;
  padding-bottom: 15px;
}

.css-19,
[data-css-19] {
  display: block;
  font-family: "Open Sans", Helvetica, Arial, sans-serif;
  font-weight: 600;
  font-size: 14px;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  color: #333333;
}

.css-20,
[data-css-20] {
  padding: 10px 15px;
  min-height: 50px;
  border-bottom: 1px solid #ecf0f1;
  cursor: pointer;
  background-color: white;
}

.css-21,
[data-css-21] {
  box-sizing: border-box;
  align-content: center;
  align-items: center;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  flex: 1 1 100%;
  cursor: pointer;
  -webkit-align-content: center;
  -webkit-box-align: center;
  -webkit-align-items: center;
  -webkit-box-orient: horizontal;
  -webkit-box-direction: normal;
  -webkit-flex-direction: row;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -webkit-flex: 1 1 100%;
}

.css-22,
[data-css-22] {
  box-sizing: border-box;
  align-content: stretch;
  align-items: stretch;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  flex: 1 1 100%;
  -webkit-align-content: stretch;
  -webkit-box-align: stretch;
  -webkit-align-items: stretch;
  -webkit-box-orient: vertical;
  -webkit-box-direction: normal;
  -webkit-flex-direction: column;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -webkit-flex: 1 1 100%;
}

.css-23,
[data-css-23] {
  box-sizing: border-box;
  flex: 0 1 auto;
  -webkit-flex: 0 1 auto;
}

.css-24,
[data-css-24] {
  width: 16px;
  height: 16px;
  stroke: #7e8c8d;
}

@media (min-width: 641px) and (max-width: 1024px), (min-width: 1025px) {
  .css-li9bei,
  [data-css-li9bei] {
    padding-left: 25px;
    padding-right: 25px;
  }
}

@media (min-width: 1025px), (min-width: 641px) and (max-width: 1024px) {
  .css-8,
  [data-css-8] {
    padding: 20px 0 0 0;
  }
}

<Connect(Connect(withRequest(Connect(ComponentWithMixpanel))))>
  <Connect(withRequest(Connect(ComponentWithMixpanel)))
    addresses={
      Array [
        Object {
          "id": "1234",
          "key": "Address#1",
        },
        Object {
          "id": "3456",
          "key": "Address#2",
        },
      ]
    }
    dispatch={[Function]}
    locale="en_US"
  >
    <Connect(ComponentWithMixpanel)
      addresses={
        Array [
          Object {
            "id": "1234",
            "key": "Address#1",
          },
          Object {
            "id": "3456",
            "key": "Address#2",
          },
        ]
      }
      createRequest={[Function]}
      dispatch={[Function]}
      locale="en_US"
    >
      <ForwardRef
        data-css-3=""
        direction="column"
        flex="flex"
      >
        <div
          data-css-6=""
          data-css-3=""
        >
          <ForwardRef
            padded="horizontal"
          >
            <ForwardRef
              data-css-7=""
              data-css-li9bei=""
              direction="column"
              flex="flex"
              onScroll={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              style={Object {}}
            >
              <div
                data-css-6=""
                data-css-7=""
                data-css-li9bei=""
                onScroll={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                style={Object {}}
              >
                <ForwardRef
                  direction="row"
                >
                  <div
                    data-css-9=""
                  >
                    <ForwardRef
                      data-css-8=""
                      flex="grow"
                    >
                      <div
                        data-css-8=""
                        data-css-10=""
                      >
                        <ForwardRef
                          data-css-4=""
                        >
                          <div
                            data-css-11=""
                            data-css-4=""
                          >
                            <div
                              data-css-12=""
                            >
                              Craftspeople
                            </div>
                          </div>
                        </ForwardRef>
                      </div>
                    </ForwardRef>
                  </div>
                </ForwardRef>
                <ForwardRef
                  data-css-13=""
                  flex="flex"
                >
                  <div
                    data-css-15=""
                    data-css-13=""
                  >
                    <ForwardRef
                      data-css-14=""
                      data-css-5=""
                      direction="row"
                      flex="flex"
                      onTransitionEnd={[Function]}
                    >
                      <div
                        data-css-14=""
                        data-css-16=""
                        data-css-5=""
                        onTransitionEnd={[Function]}
                      >
                        <ForwardRef
                          data-css-2=""
                          direction="column"
                          flex="none"
                          key="0/.0"
                        >
                          <div
                            data-css-2=""
                            data-css-17=""
                          >
                            <ForwardRef>
                              <div
                                data-css-11=""
                              >
                                <ForwardRef
                                  data-css-18=""
                                >
                                  <div
                                    data-css-18=""
                                    data-css-11=""
                                  >
                                    <ForwardRef
                                      data-css-19=""
                                    >
                                      <div
                                        data-css-19=""
                                        data-css-11=""
                                      />
                                    </ForwardRef>
                                  </div>
                                </ForwardRef>
                                <ForwardRef
                                  data-css-0=""
                                >
                                  <div
                                    data-css-0=""
                                    data-css-11=""
                                  >
                                    <ForwardRef>
                                      <ForwardRef
                                        direction="column"
                                      >
                                        <div
                                          data-css-17=""
                                        >
                                          <ForwardRef
                                            alignV="center"
                                            data-css-20=""
                                            data-e2e="1234-craftspeople-address"
                                            direction="row"
                                            flex="grow"
                                            onClick={[Function]}
                                          >
                                            <div
                                              data-css-20=""
                                              data-css-21=""
                                              data-e2e="1234-craftspeople-address"
                                              onClick={[Function]}
                                            >
                                              <ForwardRef
                                                data-css-2=""
                                                direction="column"
                                                flex="grow"
                                              >
                                                <div
                                                  data-css-22=""
                                                  data-css-2=""
                                                >
                                                  Address#1
                                                </div>
                                              </ForwardRef>
                                              <ForwardRef
                                                flex="nogrow"
                                              >
                                                <div
                                                  data-css-23=""
                                                >
                                                  <ForwardRef
                                                    alignH="center"
                                                    alignV="center"
                                                    dangerouslySetInnerHTML={
                                                      Object {
                                                        "__html": "",
                                                      }
                                                    }
                                                    data-css-24=""
                                                    data-css-1=""
                                                  >
                                                    <div
                                                      dangerouslySetInnerHTML={
                                                        Object {
                                                          "__html": "",
                                                        }
                                                      }
                                                      data-css-24=""
                                                      data-css-1=""
                                                      data-css-11=""
                                                    />
                                                  </ForwardRef>
                                                </div>
                                              </ForwardRef>
                                            </div>
                                          </ForwardRef>
                                          <ForwardRef
                                            alignV="center"
                                            data-css-20=""
                                            data-e2e="3456-craftspeople-address"
                                            direction="row"
                                            flex="grow"
                                            onClick={[Function]}
                                          >
                                            <div
                                              data-css-20=""
                                              data-css-21=""
                                              data-e2e="3456-craftspeople-address"
                                              onClick={[Function]}
                                            >
                                              <ForwardRef
                                                data-css-2=""
                                                direction="column"
                                                flex="grow"
                                              >
                                                <div
                                                  data-css-22=""
                                                  data-css-2=""
                                                >
                                                  Address#2
                                                </div>
                                              </ForwardRef>
                                              <ForwardRef
                                                flex="nogrow"
                                              >
                                                <div
                                                  data-css-23=""
                                                >
                                                  <ForwardRef
                                                    alignH="center"
                                                    alignV="center"
                                                    dangerouslySetInnerHTML={
                                                      Object {
                                                        "__html": "",
                                                      }
                                                    }
                                                    data-css-24=""
                                                    data-css-1=""
                                                  >
                                                    <div
                                                      dangerouslySetInnerHTML={
                                                        Object {
                                                          "__html": "",
                                                        }
                                                      }
                                                      data-css-24=""
                                                      data-css-1=""
                                                      data-css-11=""
                                                    />
                                                  </ForwardRef>
                                                </div>
                                              </ForwardRef>
                                            </div>
                                          </ForwardRef>
                                        </div>
                                      </ForwardRef>
                                    </ForwardRef>
                                  </div>
                                </ForwardRef>
                              </div>
                            </ForwardRef>
                          </div>
                        </ForwardRef>
                      </div>
                    </ForwardRef>
                  </div>
                </ForwardRef>
              </div>
            </ForwardRef>
          </ForwardRef>
        </div>
      </ForwardRef>
    </Connect(ComponentWithMixpanel)>
  </Connect(withRequest(Connect(ComponentWithMixpanel)))>
</Connect(Connect(withRequest(Connect(ComponentWithMixpanel))))>
`;
